Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP1Connections.swift

```diff
@@ -0,0 +1,602 @@
+            }
+        }
+
+        // MARK: Connection close/removal
+
+        mutating func closeConnectionIfIdle(_ connectionID: Connection.ID) -> Connection? {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                // because of a race this connection (connection close runs against trigger of timeout)
+                // was already removed from the state machine.
+                return .none
+            }
+
+            guard self.connections[index].isAvailable else {
+                // connection is not available anymore, we may have just leased it for a request
+                return nil
+            }
+
+            return self.closeConnection(at: index)
+        }
+
+        // MARK: Connection failure
+
+        /// A connection's use. Did it serve in the pool or was it specialized for an `EventLoop`?
+        enum ConnectionUse {
+            case generalPurpose
+            case eventLoop(EventLoop)
+        }
+
+        /// Information around the failed/closed connection.
+        struct FailedConnectionContext {
+            /// The eventLoop the connection ran on.
+            var eventLoop: EventLoop
+            /// The failed connection's use.
+            var use: ConnectionUse
+            /// Connections that we start up for this use-case
+            var connectionsStartingForUseCase: Int
+        }
+
+        enum FailedConnectionAction {
+            /// Creates a new connection, that needs to be started
+            case replace
+            /// Removes the old connection.
+            case remove
+        }
+
+        /// Remove or replace a connection.
+        ///
+        /// Call this function after the connection start backoff is done, or a live connection is closed.
+        ///
+        /// - Parameters:
+        ///   - connectionID: The failed connection's ID
+        ///   - closure: A closure in which a decision is made how to progress with this connection spot
+        ///
+        /// - Returns: A new connectionID and eventLoop to create a new connection, if the action is `.replace`
+        mutating func failConnection(
+            _ connectionID: Connection.ID,
+            _ closure: (FailedConnectionContext) -> (FailedConnectionAction)
+        ) -> (Connection.ID, EventLoop)? {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to create a new connection that we know nothing about?")
+            }
+
+            let use: ConnectionUse
+            let eventLoop = self.connections[index].eventLoop
+            let starting: Int
+            if index < self.overflowIndex {
+                use = .generalPurpose
+                // we need to subtract 1, since we dont want this connection to count
+                starting = self.startingGeneralPurposeConnections - 1
```

Comment:
Is this right? If we're a live connection that's been closed then we won't be in the starting state.